File: markdownMapper.py
import os
import re
from collections import defaultdict

def generate_link(file, header=None):
    if header:
        header_text = header.strip("#").strip()
        header_id = header_text.lower().replace(" ", "-")
    else:
        header_text = file[:-3]  # Remove file extension
        header_id = ""
    link = f"[{header_text}]({file}#{header_id})"
    return link

def generate_toc(file):
    toc = defaultdict(list)
    main_heading = None
    with open(file, "r") as md_file:
        lines = md_file.readlines()
        for line in lines:
            if line.startswith("#"):
                level = line.count("#", 0, 3)  # only consider levels up to ###
                link = generate_link(file, line)
                if level == 1:
                    main_heading = link
                elif main_heading:  # Only add subheadings if there is a main heading
                    toc[main_heading].append((level, link))

    # Add the file name as a main heading if there is no main heading
    if not main_heading:
        main_heading = generate_link(file)
        toc[main_heading] = []

    return toc

def main():
    toc_filename = "table_of_contents.md"
    toc_dict = defaultdict(list)

    for file in os.listdir("."):
        if file.endswith(".md") and file != toc_filename:
            toc_file = generate_toc(file)
            for main_heading, subheadings in toc_file.items():
                toc_dict[main_heading].extend(subheadings)

    sorted_toc = sorted(toc_dict.items(), key=lambda x: x[0].lower())

    with open(toc_filename, "w") as toc_file:
        toc_file.write("Table of Contents:\n")
        for main_heading, subheadings in sorted_toc:
            toc_file.write(f"- {main_heading}\n")
            for level, link in subheadings:
                indent = "    " * (level - 1)
                toc_file.write(f"{indent}- {link}\n")

if __name__ == "__main__":
    main()

File: markdownMapper.go
package main

import (
	"bufio"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

type Entry struct {
	level int
	text  string
}

func generateLink(file string, header string) string {
	headerText := strings.TrimSpace(strings.TrimLeft(header, "#"))
	headerID := strings.Replace(strings.ToLower(headerText), " ", "-", -1)
	link := fmt.Sprintf("[%s](./%s#%s)", headerText, file, headerID)
	return link
}

func generateTOC(file string) map[string][]Entry {
	toc := make(map[string][]Entry)
	f, _ := os.Open(file)
	scanner := bufio.NewScanner(f)

	headerRegex := regexp.MustCompile(`^(#{1,3})\s+`)
	mainHeading := ""

	for scanner.Scan() {
		line := scanner.Text()
		if headerRegex.MatchString(line) {
			level := len(headerRegex.FindStringSubmatch(line)[1])
			link := generateLink(file, line)
			indent := strings.Repeat("    ", level-1)
			if level == 1 {
				mainHeading = link
			} else if mainHeading != "" {
				toc[mainHeading] = append(toc[mainHeading], Entry{level: level, text: indent + "- " + link})
			}
		}
	}

	if mainHeading == "" {
		mainHeading = fmt.Sprintf("[%s](./%s)", strings.TrimSuffix(file, filepath.Ext(file)), file)
		toc[mainHeading] = []Entry{}
	}

	return toc
}

func main() {
	tocFilename := "table_of_contents.md"
	tocMap := make(map[string][]Entry)

	files, _ := ioutil.ReadDir(".")
	for _, f := range files {
		file := f.Name()
		if strings.HasSuffix(file, ".md") && file != tocFilename {
			fileToc := generateTOC(file)
			for mainHeading, subheadings := range fileToc {
				tocMap[mainHeading] = append(tocMap[mainHeading], subheadings...)
			}
		}
	}

	sortedToc := make([]string, 0, len(tocMap))
	for mainHeading := range tocMap {
		sortedToc = append(sortedToc, mainHeading)
	}
	sort.Slice(sortedToc, func(i, j int) bool {
		return strings.ToLower(sortedToc[i]) < strings.ToLower(sortedToc[j])
	})

	tocFile, _ := os.Create(tocFilename)
	defer tocFile.Close()

	tocWriter := bufio.NewWriter(tocFile)
	tocWriter.WriteString("Table of Contents:\n")
	for _, mainHeading := range sortedToc {
		tocWriter.WriteString("- " + mainHeading + "\n")
		for _, entry := range tocMap[mainHeading] {
			tocWriter.WriteString(entry.text + "\n")
		}
	}
	tocWriter.Flush()
}
